### PR TITLE
fix various memory leaks

### PR DESCRIPTION
**Pull request by sebsura**
#### Please check

- [X] Short description and the purpose of this PR is present _above this paragraph_
- [X] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)
- [Boy Scout Rule](https://docs.bareos.org/DeveloperGuide/generaldevel.html#boy-scout-rule)

### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
- [x] Required backport PRs have been created

##### Source code quality
- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR

##### Tests
- [x] Decision taken that a test is required (if not, then remove this paragraph)
- [x] The choice of the type of test (unit test or systemtest) is reasonable
- [x] Testname matches exactly what is being tested
- [x] On a fail, output of the test leads quickly to the origin of the fault

### CODE REVIEW

**Review by arogge — CHANGES_REQUESTED** [review]
That is a huge improvement! How did you find all of these?
I have a few remarks. Nothing critical though.

**Comment by sebsura** [comment]
I gathered all the logs created by asan with our test suite.

File: core/src/dird/dbcheck_utils.cc [hunk 1]
```diff
@@ -105,6 +105,14 @@ int MakeIdList(BareosDb* db, const char* query, ID_LIST* id_list)
...
 }
 
+void FreeIdList(s_id_ctx* id_list)
+{
+  id_list->num_ids = 0;
+  id_list->num_del = 0;
+  id_list->tot_ids = 0;
+  free(id_list->Id);
+}
+
```
[the remark this answers is not in the document]
Comment (sebsura) [reply core/src/dird/dbcheck_utils.cc:115]: Just noticed that they are in fact used by `dbcheck.cc`, just not cleaned up.  I guess we do not have a test that currently tests this code path.  Ill see if I can add one test.

File: core/src/stored/acquire.cc [hunk 2]
```diff
@@ -172,14 +171,14 @@ bool AcquireDeviceForRead(DeviceControlRecord* dcr)
     jcr->sd_impl->reserve_msgs = new alist<const char*>(10, not_owned_by_alist);
     rctx.any_drive = true;
     rctx.device_name = vol->device;
-    store = new DirectorStorage;
-    memset(store, 0, sizeof(DirectorStorage));
+    auto store = std::make_unique<DirectorStorage>();
+    memset(store.get(), 0, sizeof(DirectorStorage));
```
[the remark this answers is not in the document]
Comment (sebsura) [reply core/src/stored/acquire.cc:175]: done

File: core/src/stored/acquire.cc [hunk 3]
```diff
@@ -172,14 +171,14 @@ bool AcquireDeviceForRead(DeviceControlRecord* dcr)
...
     store->name[0] = 0; /* No dir name */
     bstrncpy(store->media_type, vol->MediaType, sizeof(store->media_type));
     bstrncpy(store->pool_name, dcr->pool_name, sizeof(store->pool_name));
     bstrncpy(store->pool_type, dcr->pool_type, sizeof(store->pool_type));
     store->append = false;
```
[the remark this answers is not in the document]
Comment (sebsura) [reply core/src/stored/acquire.cc:180]: done

File: core/src/stored/acquire.cc [hunk 4]
```diff
@@ -172,14 +171,14 @@ bool AcquireDeviceForRead(DeviceControlRecord* dcr)
...
-    rctx.store = store;
+    rctx.store = store.get();
```
[the remark this answers is not in the document]
Comment (sebsura) [reply core/src/stored/acquire.cc:181]: done

File: core/src/dird/dbcheck_utils.h [hunk 5]
```diff
@@ -68,6 +68,7 @@ int NameListHandler(void* ctx, int num_fields, char** row);
 int MakeNameList(BareosDb* db, const char* query, s_name_ctx* name_list);
 void PrintNameList(s_name_ctx* name_list);
 void FreeNameList(s_name_ctx* name_list);
+void FreeIdList(s_id_ctx* id_list);
```
[the remark this answers is not in the document]
Comment (sebsura) [reply core/src/dird/dbcheck_utils.h:71]: moved it next to the other idlist functions

**Review by arogge — APPROVED** [review]
The code looks great.
However, the new dbcheck test fails everywhere, so you'll have to take another look.

**Comment by arogge** [comment]
Do you think we should backport some/all of these into 23?

**Comment by sebsura** [comment]
Ill try to backport it and ill see what works.

**Pull request merged**